File: frontend/maquette/harness/season_grab.py
"""R125 — « Récupérer cette saison » really takes the season (B-301).

THE SEASON MATRIX PRINTS A HOLE AND OFFERS NOTHING. `panel-seasons.tsx` draws
one cell per episode and marks a season whose owned count is short of what
aired — « Saison 3 · 6/7 · 1 manquant ». The interface SHOWS what is wanted and
gives no way to ask for it, while
`POST /api/acquisition/follows/{followedId}/seasons/{season}/grab` goes uncalled.
DOIT-3 is « agir là où l'on observe », and this is the clearest place in the
application where one cannot.

⚠ THE HOLE IS NOT A `to_grab` CELL, whatever a brief may say. Measured: the
fixture contains **no `to_grab` episode cell anywhere**, because
`epState` colours a missing episode by the FOLLOW's status and the one follow
with a hole is `pending`. The subject this rule walks is the only one that
exists — « Silo », season 3, six of seven aired episodes held — and it is chosen
from `window.SEASONS` (owned < aired) BEFORE a finger moves, so a failure is
about the interface and never about the walk.

THIS RULE IS WRITTEN BEFORE THE VERB EXISTS and is RED against `main` with no
mutation needed — the strongest form of « seen red first » this repository asks
for: a repair held by nothing comes back with its sign turned round, which is
what `data-take` cost before a rule read it.

WHAT IT READS, and each fails differently:

  1. A SEASON WITH A HOLE OFFERS THE VERB, reached by a FINGER. Not by the
     seam: `window.__panel.produce` opens the panel without walking the path
     that raises it, so a row that had lost its `data-panel` would be invisible
     here while every hold below stayed green. That is the vacuity R124 was
     repaired for, and it is not reproduced.
  2. THE OPERATION IS CALLED, read on the NETWORK. A hold reading the screen
     alone passes a build that toasted and sent nothing — which is exactly what
     « Récupérer maintenant » did for a whole wave under a green gate (B-309).
     The request's own path is read back, so a call to the WRONG season is a
     failure and not a pass.
  3. THE STATE MOVES. The follow leaves the status it had. A message is a
     message and can be right about nothing (NE-DOIT-PAS-1).
  4. NO ERROR IS RAISED — B-309's signature, and the reason a hold reading only
     the counts could pass a build that threw on the way.
  5. UNDER A BUSY PIPELINE the ask is still accepted, says it is QUEUED, and is
     never answered 409 nor dressed as « occupé » (DOIT-4, NE-DOIT-PAS-3, §20).

WHY THE HIT TEST IS COPIED HERE rather than shared: `busy.py` carries the same
helper, and the natural home for a second copy is `common.py` — which cannot be
opened without taking B-325 (no rule can be pointed at a build; `PROTOTYPE` is
hard-coded with no override). A third copy is where that move stops being
optional; this is the second, and it is said out loud so the next reader decides
rather than discovers.
"""
import asyncio
import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from common import ACTED, Journal, PANEL_IN, PANEL_OUT, SETTLED, open_page

from playwright.async_api import async_playwright

# WHERE A FOLLOW IS DRAWN WITH A ROW A FINGER CAN REACH. The acquisitions page
# lists what is in FLIGHT, so a followed medium not currently being acquired has
# no row there at all — the correction R124 paid for, taken here from the start.
FOLLOWS_STATE = "acq-follows-list"

# The operation, as the CONTRACT names it. It is read by operationId rather than
# by URL, so a path parameter spelled differently cannot turn a real call into a
# silent miss.
GRAB_OPERATION = "grabSeasonForFollow"

# WHAT THE LAYER ANSWERED, and why it is not read on the network. The mock layer
# REPLACES `globalThis.fetch`, so a mocked call reaches no network at all:
# measured, a verb that demonstrably ran — the follow moved from `pending` to
# `acquiring` — produced ZERO Playwright request events. A hold listening on
# `page.on("request")` or `page.on("response")` for one of these operations is
# therefore green whatever the interface does, which is a guard certifying what
# it cannot see. The layer records what it answered; this reads that.
ANSWERED = "()=>(window.__mocks?.answered?.() || [])"

# THE WORDS A REFUSAL WEARS. « occupé » is the constitution's own; the others
# are what the same refusal reads like when it is dressed differently.
REFUSALS = ("occupé", "occupee", "occupée", "déjà en cours", "réessayez plus tard")

FOLLOWS = """()=>(window.__followActions?.all() || []).map(
  (one) => ({t: one.t, st: one.st}))"""

# WHICH MEDIUM HAS A SEASON WITH A HOLE, decided from the DATA before a finger
# moves. `window.SEASONS` is `[number, aired, owned]` per season; a hole is
# `owned < aired` — an episode that HAS aired and is not held, which is exactly
# what a season grab is for. An unaired episode is not a hole: it is not out yet.
# The follow must also be drawn on this page, or no finger could reach it.
THE_MEDIUM_WITH_A_HOLE = """()=>{
  const drawn = [...document.querySelectorAll('[data-panel]')].map(
    (one) => one.dataset.panel);
  const reachable = (title) => drawn.some(
    (seen) => seen === title || seen.endsWith(":" + title));
  for (const follow of (window.__followActions?.all() || [])) {
    if (!reachable(follow.t)) continue;
    for (const [number, aired, owned] of (window.SEASONS[follow.t] || [])) {
      if ((owned || 0) < (aired || 0))
        return {title: follow.t, season: number, aired, owned};
    }
  }
  return null;}"""

# WHAT THE INTERFACE SAID, read through the door the message host publishes FOR
# a rule. It is NOT `#toast`: that element is the dying engine's, and the
# message layer is React — measured, `#toast` stayed EMPTY through a verb whose
# message was on screen the whole time. A hold reading the old element would
# have been green for « never says occupé » while saying nothing at all, and red
# for « says En file » while the interface said it perfectly.
SAID = """()=>{
  const held = window.__toast?.read?.();
  const message = held && held.message ? held.message.message || '' : '';
  const page = [...document.querySelectorAll('#view')]
    .map((node) => node.textContent || '').join(' ');
  return {said: message, shown: !!(held && held.shown), page: page};}"""

# PUTS THE LAYER TO WORK, through the operation the application itself calls.
# `runPipeline` moves the mock's own pipeline state to running (or to queued
# when one is already going), and that is the state the three verbs read to
# decide whether an ask waits. It answers what the state became, so a walk
# cannot believe it succeeded.
RUN_THE_PIPELINE = """async()=>{
  const answer = await window.fetch("/api/pipeline/run", {method: "POST"});
  const body = await answer.json().catch(() => ({}));
  return body.state || "";}"""

# WHAT COVERS THE ROW IS READ, NEVER ASSUMED. `elementFromPoint` at the row's
# own centre answers what a tap there would actually hit, which is a different
# question from « is the node in the tree ».
AIM_AT_THE_ROW = """(title)=>{
  const row = [...document.querySelectorAll('[data-panel]')].find(
    (one) => one.dataset.panel === title
          || one.dataset.panel.endsWith(":" + title));
  if (!row) return {found: false};
  const box = row.getBoundingClientRect();
  const x = box.left + box.width / 2;
  const y = box.top + box.height / 2;
  const hit = document.elementFromPoint(x, y);
  return {found: true, x, y,
          reachable: !!hit && (hit === row || row.contains(hit)),
          covering: hit === null ? "nothing" :
            (hit.tagName + (hit.className ? "." + String(hit.className).split(" ")[0] : ""))};}"""

# The greeting toast sits over the surface (B-317) and a tap under it lands on
# the toast. No gesture dismisses it, so the element is emptied — said plainly,
# because this walk therefore does not prove the toast can be dismissed, only
# that what is under it is reachable once it is gone.
EMPTY_THE_TOAST = """()=>{const one = document.querySelector('#toast');
  if (!one) return false;
  one.textContent = "";
  one.className = "";
  one.removeAttribute("data-open");
  return true;}"""

# A SEASON THAT HAS A HOLE, and the verb offered on its row.
#
# THE HOLE IS `season/missing`, NOT A `to_grab` CELL, and that correction was
# paid for by a first version of this rule. It looked for an episode cell
# classed `to_grab`, on the brief's own words — « a season printed to_grab » —
# and went red saying « none offered a hole » across all twelve follows. The
# reason was not the missing verb: **the fixture contains no `to_grab` cell at
# all.** `epState` colours a missing episode by the FOLLOW's status, so a follow
# that is `pending` draws its holes `pending`, and every other follow in the
# fixture is `up_to_date` with nothing missing. A rule red for a reason
# unrelated to what it measures is a rule that would still be red after the
# repair, and would then be "fixed" into a green that never read the verb.
#
# What the panel really prints over a hole is `[data-part="season/missing"]` —
# the « N manquants » mark, drawn exactly when owned < aired. That is what a
# season grab is FOR, whatever lifecycle colour the cells happen to wear.
SEASON_WITH_A_HOLE = """()=>{
  const seasons = [...document.querySelectorAll('#sheetin [data-part="season"]')];
  for (const season of seasons) {
    const hole = season.querySelector('[data-part="season/missing"]');
    if (!hole) continue;
    const act = season.querySelector('[data-part="season/grab"]');
    const box = act ? act.getBoundingClientRect() : null;
    const hit = box ? document.elementFromPoint(
      box.left + box.width / 2, box.top + box.height / 2) : null;
    return {
      label: ((season.querySelector("summary") || {}).textContent || "").trim(),
      wanted: (hole.textContent || "").trim(),
      offered: !!act,
      value: act ? (act.dataset.grabSeason || "") : "",
      x: box ? box.left + box.width / 2 : 0,
      y: box ? box.top + box.height / 2 : 0,
      reachable: !!hit && (hit === act || (act && act.contains(hit)))};
  }
  return {label: "", wanted: "", offered: false, value: "", reachable: false};}"""


async def raise_by_finger(page, title):
    """Raises a row's panel with a hit-tested tap, and says what it hit.

    Args:
        page: The page.
        title: The subject the row's `data-panel` names.

    Returns:
        The aim's own reading, with `tapped` saying whether a finger went down.
    """
    await page.evaluate(EMPTY_THE_TOAST)
    aim = await page.evaluate(AIM_AT_THE_ROW, title)
    if aim["found"] and aim["reachable"]:
        await page.touchscreen.tap(aim["x"], aim["y"])
        aim["tapped"] = True
    else:
        aim["tapped"] = False
    return aim


async def open_a_season_panel(page, journal, when, put_to_work=False):
    """Drives the follows list and raises the panel of the one medium with a hole.

    THE SUBJECT IS CHOSEN FROM THE DATA, AND ONE FINGER GOES TO IT. An earlier
    version tapped every row in turn looking for a hole, and measured itself
    instead of the interface: the FIRST panel it opened covered the rows behind
    it, so `elementFromPoint` at the next row's centre hit the panel, eleven
    taps never landed, and the rule reported « none printing a season with a
    hole » over twelve media it had never actually looked at. Reading
    `window.SEASONS` for a season where owned < aired names the subject before
    any gesture, so the walk is one tap and the failure — if it comes — is about
    the interface rather than about the walk.

    THE MACHINE IS PUT TO WORK AFTER THE STATE IS DRIVEN, never before, and
    that ordering is the whole of the busy half. `window.__go` RE-SEEDS the mock
    layer, so a pipeline set running before it is idle again by the time the act
    lands: the walk then measured a clause it had switched off, and the hold
    saying « the pipeline really is busy » was green over a reading already
    discarded. It is checked here, after the state, where it still means
    something.

    Args:
        page: The page.
        journal: Where the holds are recorded.
        when: What this half is measuring, for the hold's own words.
        put_to_work: Whether to start the pipeline once the state is driven.

    Returns:
        The medium's title, and the season reading, or ("", None) when the
        fixture holds no season with a hole, or no finger can raise its panel.
    """
    await page.evaluate("(id)=>window.__go(id)", FOLLOWS_STATE)
    await page.wait_for_timeout(SETTLED)
    if put_to_work:
        started = await page.evaluate(RUN_THE_PIPELINE)
        await page.evaluate("""()=>window.__store.write({pipe: "running"})""")
        await page.wait_for_timeout(SETTLED)
        journal.check(
            "the LAYER really has the pipeline busy at the moment of the act — "
            "the half that decides whether an ask is queued",
            started in ("running", "queued"), str(started))
        journal.check(
            "and the interface is drawing it busy too, so this half measures "
            "the clause and not the verb",
            await page.evaluate("()=>window.__store.read().state.pipe") == "running")
    subject = await page.evaluate(THE_MEDIUM_WITH_A_HOLE)
    journal.check(
        f"the fixture really holds a season with a hole, so this walk has a "
        f"subject — {when}",
        bool(subject), str(subject) or "no follow has owned < aired in window.SEASONS")
    if not subject:
        return "", None
    title = subject["title"]
    aim = await raise_by_finger(page, title)
    journal.check(
        f"« {title} » has a row a finger can raise its panel from",
        aim["tapped"],
        f"found={aim['found']} reachable={aim.get('reachable')} "
        f"at ({aim.get('x')}, {aim.get('y')}) hits {aim.get('covering')}")
    if not aim["tapped"]:
        return "", None
    await page.wait_for_timeout(PANEL_IN)
    season = await page.evaluate(SEASON_WITH_A_HOLE)
    if not season["label"]:
        journal.check(
            f"and the panel it raises PRINTS that hole — season "
            f"{subject['season']}, {subject['owned']} of {subject['aired']}",
            False, "the panel drew no season carrying a missing mark")
        return "", None
    return title, season


async def main():
    journal = Journal("R125 — « Récupérer cette saison » really takes the season")
    async with async_playwright() as playwright:
        browser = await playwright.chromium.launch(channel="chrome")
        context, page = await open_page(browser)
        errors: list[str] = []
        page.on("pageerror", lambda error: errors.append(str(error)))

        async def grabs_since(mark):
            """Every season-grab call the layer answered after a mark.

            Args:
                mark: How many calls had been answered before the act.

            Returns:
                The calls, as the layer recorded them.
            """
            every = await page.evaluate(ANSWERED)
            return [call for call in every[mark:]
                    if call["operationId"] == GRAB_OPERATION]

        async def answered_so_far():
            """How many calls the layer has answered, as a mark."""
            return len(await page.evaluate(ANSWERED))

        # ── THE VERB, ON AN IDLE PIPELINE ──────────────────────────────────
        title, season = await open_a_season_panel(page, journal, "the pipeline idle")
        if season is not None:
            journal.check(
                f"« {title} » prints a season with a hole — {season['label']}",
                bool(season["wanted"]), season["wanted"] or "no mark drawn")
            journal.check(
                "and that season OFFERS to be taken, on a button a finger reaches "
                "(DOIT-3, B-301)",
                season["offered"] and season["reachable"],
                f"offered={season['offered']} reachable={season['reachable']} "
                f"value={season['value']!r}")
            was = await page.evaluate(
                "(t)=>(window.__followActions?.all() || []).find((one) => one.t === t)?.st",
                title)
            errors.clear()
            mark = await answered_so_far()
            if season["offered"] and season["reachable"]:
                await page.touchscreen.tap(season["x"], season["y"])
                await page.wait_for_timeout(ACTED)
            # THE OPERATION IS CALLED, and the call is read on the NETWORK. A
            # hold reading the screen alone passes a build that toasted and
            # sent nothing.
            grabs = await grabs_since(mark)
            journal.check(
                "the season grab OPERATION is called — read on what the LAYER "
                "answered, never on the screen (B-301)",
                len(grabs) == 1, f"{len(grabs)} call(s): {grabs[:2]}")
            # AND FOR THE SEASON THE FINGER WAS ON. A call to the wrong season
            # is a call, and a hold counting calls alone would take it.
            wanted_season = season["value"].split("|")[-1] if season["value"] else ""
            journal.check(
                "and it names the season the finger was on, not another",
                bool(wanted_season) and any(
                    call["path"].endswith(f"/seasons/{wanted_season}/grab")
                    for call in grabs),
                f"season {wanted_season!r} against {[c['path'] for c in grabs]}")
            journal.check(
                "tapping it raises no error",
                not errors, str(errors))
            now = await page.evaluate(
                "(t)=>(window.__followActions?.all() || []).find((one) => one.t === t)?.st",
                title)
            journal.check(
                "and the STATE moves — the act lands, it is not a message about "
                "an act (NE-DOIT-PAS-1)",
                now is not None and now != was, f"{title}: {was} → {now}")

        # ── AND UNDER A BUSY PIPELINE, where the clause bites ──────────────
        #
        # THE LAYER IS PUT TO WORK THROUGH ITS OWN OPERATION, and both truths
        # are set because there are two. The engine's `pipe` field drives the
        # DRAWING; the mock layer's `pipelineState` is what decides `queued` in
        # the answer. A walk that wrote only the store would leave the layer
        # idle and read a queued flag nobody set — the hold would be green over
        # nothing. `runPipeline` is the operation the application itself calls,
        # so this drives the layer rather than reaching past it.
        # THE FIRST HALF'S PANEL IS CLOSED FIRST, and its scrim with it. Left
        # open, the panel covers the follows list and `elementFromPoint` at the
        # row's centre answers `DIV.scrim` — the second half then failed on the
        # tap it never landed rather than on the clause it exists to read, which
        # is a rule measuring its own leftovers.
        await page.evaluate("()=>window.__panel.close()")
        await page.wait_for_timeout(PANEL_OUT + SETTLED)
        busy_title, busy_season = await open_a_season_panel(
            page, journal, "the pipeline busy", put_to_work=True)
        if busy_season is not None:
            busy_mark = await answered_so_far()
            errors.clear()
            if busy_season["offered"] and busy_season["reachable"]:
                await page.touchscreen.tap(busy_season["x"], busy_season["y"])
                await page.wait_for_timeout(ACTED)
            busy_grabs = await grabs_since(busy_mark)
            journal.check(
                f"« {busy_title} »'s season can still be asked for while the "
                "pipeline runs — the ask is ACCEPTED (NE-DOIT-PAS-3)",
                len(busy_grabs) == 1, f"{len(busy_grabs)} call(s)")
            spoken = await page.evaluate(SAID)
            said = spoken["said"].lower()
            journal.check(
                "and the interface says it is IN FILE — DOIT-4's own words, "
                "« En file — pipeline en cours »",
                spoken["shown"] and "en file" in said,
                f"shown={spoken['shown']} said={spoken['said']!r}")
            everywhere = (said + " " + spoken["page"]).lower()
            journal.check(
                "never « occupé », nor any other dress of the same refusal",
                not any(word in everywhere for word in REFUSALS),
                next((word for word in REFUSALS if word in everywhere), ""))

        # READ ON WHAT THE LAYER ANSWERED, for the reason above: a `page.on
        # ("response")` listener cannot see a status this layer never put on a
        # wire, so a 409 hold written that way is green by construction.
        every = await page.evaluate(ANSWERED)
        refused = [call for call in every if call["status"] == 409]
        journal.check(
            "and no ask anywhere was answered 409 (NE-DOIT-PAS-3)",
            not refused, str(refused[:3]))

        await context.close()
        await browser.close()
    journal.summary()


if __name__ == "__main__":
    asyncio.run(main())

File: frontend/maquette/design/src/features/media/season-grab.ts
// « Récupérer cette saison » — the verb a season with a hole owes (B-301).
//
// THE SEASON MATRIX SHOWED THE HOLE AND OFFERED NOTHING. It draws « Saison 3 ·
// 6/7 · 1 manquant » and there was no way to ask for the missing episode.
// DOIT-3 is « agir là où l'on observe »; this is where one observes it.
//
// IT LIVES IN `features/media/` BECAUSE THAT IS WHERE THE PANEL IS DRAWN, and
// invariant 7 is absolute: the media feature may not import
// `features/acquisition/`. What it does instead is call the OPERATION — a
// mutation on the contract is not a feature import. `features/media/queries.ts`
// is this feature's own door to the layer, and this is its mutating half.
//
// A FILE OF ITS OWN beside the block it serves, never a growth of
// `panel-seasons.tsx`: a verb is behaviour, the block is drawing, and invariant
// 6 asks for the cut before the ceiling rather than after it.
import type { QueryClient } from "@tanstack/react-query";
import { useSyncExternalStore } from "react";
import { markSeasonQueued } from "./queued-seasons";
import i18next from "i18next";
import { HELD, send } from "../../lib/query-client";

/** What the operation answers, as the contract declares it. */
type SeasonGrab = {
  season: number;
  absorbedCount: number;
  queued: boolean;
  runUid: string | null;
  newlyFollowed: boolean;
};

/**
 * Asks for one season of one follow.
 *
 * WHAT IT SAYS DEPENDS ON WHAT CAME BACK, and that is DOIT-4 rather than a
 * nicety. An ask that arrives while the pipeline is working is QUEUED — the
 * interface says « En file — pipeline en cours » and never « occupé », and it
 * never shows the 409 the backend answers for that case (NE-DOIT-PAS-3, §20).
 * An ask that starts now says so. **A toast is written from the ANSWER, never
 * before it**: a sentence composed optimistically is a sentence that can be
 * right about nothing, which is NE-DOIT-PAS-1 and the defect `data-take` wore
 * for a whole wave.
 *
 * OFFLINE IS NOT A FAILURE. `send` answers `HELD` when the mutation is queued
 * for a network that is not there; saying « asked for » then would be a lie and
 * saying « failed » would be a worse one, so the interface says it is held.
 *
 * Args:
 *     title: The follow, as this interface knows it — by TITLE. The backend
 *         wants a rowid; that is a demand on the register (§ 2b), not something
 *         to invent here.
 *     season: The season number, 1-based.
 *
 * Returns:
 *     Whether the ask is WAITING on the pipeline rather than under way. The
 *     caller records it so the season can go on saying so after the message
 *     has gone (DOIT-4's visible half): a sentence shown for four seconds
 *     tells the operator who was looking, and nobody else.
 */
/**
 * Asks for a season and settles every surface that shows it.
 *
 * WHY IT EXISTS RATHER THAN LIVING IN A HANDLER. Two surfaces draw the same
 * season with the same hole — the follow panel and the media sheet's own list —
 * and the queued mark was deliberately put on BOTH. An action drawn on one of
 * them with its behaviour written inline is an action the other cannot offer
 * without copying four steps, and a copied sequence is one that stops agreeing
 * the first time either end moves.
 *
 * THE FOUR STEPS ARE ONE UNIT: ask, remember the wait so the season goes on
 * saying so after the message has gone, re-read the follows the ask moved, and
 * put the open panel back from what came. The last one is the whole of it —
 * without it the surface the operator pressed is the only one that does not
 * know, which is where he was looking.
 *
 * Args:
 *     client: The cache both surfaces read.
 *     title: The followed medium.
 *     season: The season number, 1-based.
 */
export async function askForSeason(
  client: QueryClient,
  title: string,
  season: number,
): Promise<void> {
  const waiting = await grabSeason(title, season);
  if (waiting) markSeasonQueued(client, title, season);
  await client.refetchQueries({ queryKey: ["/api/acquisition/followed"] });
  window.__panel?.redraw();
}

/* THE ASKS IN FLIGHT, keyed `title|season`. Module state and not a React ref:
   the same season is reachable from two surfaces, and a guard held by one
   component would not see a press on the other. */
const inFlight = new Set<string>();

/* AND WHAT IS IN FLIGHT IS DRAWN. The guard below answers a second press with
   silence, and the operator saw a button that had taken nothing: with the answer
   held back 2.5 s the pressed act read no state at all, the same text at full
   opacity, while three more presses did nothing visible. So the asks in flight
   are published, and both surfaces draw their act as TAKEN while its ask is out.

   A SNAPSHOT REPLACED ON EVERY CHANGE, never the live set: the subscription
   compares what it is handed, so a set mutated in place would report no change
   and a fresh copy on every read would report one on every render. */
let flightSnapshot: ReadonlySet<string> = new Set();
const flightListeners = new Set<() => void>();

function announceFlight(): void {
  flightSnapshot = new Set(inFlight);
  for (const listener of flightListeners) listener();
}

/**
 * The season asks in flight, keyed `title|season`, for a surface drawing acts.
 *
 * Not the store, for `app/message-presence.ts`'s measured reason: a store write
 * re-renders every page, and a node replaced between a press and its click
 * loses the click (B-247).
 *
 * Returns:
 *     The asks under way right now.
 */
export function useAskedInFlight(): ReadonlySet<string> {
  return useSyncExternalStore(
    (onChange) => {
      flightListeners.add(onChange);
      return () => flightListeners.delete(onChange);
    },
    () => flightSnapshot,
  );
}

export async function grabSeason(title: string, season: number): Promise<boolean> {
  const say = (key: string, values?: Record<string, unknown>) =>
    i18next.t(`verbs.media.${key}`, values ?? {});
  // ONE ASK AT A TIME PER SEASON. Two presses with no settle between them sent
  // two identical requests and produced ONE message, because the second answer
  // replaced the first — so the interface asked twice and said so once. A
  // second press while the first is in flight is not a second intention: it is
  // the same one, made again because nothing had visibly happened yet. It is
  // answered with silence rather than a refusal, because a refusal would say
  // « occupé » to a legitimate act, which is the clause NE-DOIT-PAS-3 refuses.
  //
  // KEYED BY TITLE AND SEASON, never by title alone: asking for season 2 while
  // season 3 is in flight is a different ask and must land.
  const asked = `${title}|${season}`;
  if (inFlight.has(asked)) return false;
  inFlight.add(asked);
  announceFlight();
  try {
    const answered = await send<SeasonGrab>(
      "POST",
      `/api/acquisition/follows/${encodeURIComponent(title)}/seasons/${season}/grab`,
    );
    if (answered === HELD) {
      window.__toast?.show({ message: say("seasonHeld", { season }) });
      return false;
    }
    const grab = answered as SeasonGrab | undefined;
    // THE SENTENCE IS CHOSEN, NOT PLURALISED WITH A PARENTHESIS. « 1
    // épisode(s) » makes the reader do the grammar, and this file's own
    // neighbour — the deck's « 1 suggestion de plus » beside « N suggestions de
    // plus » — already writes each number's sentence out. Three and not two,
    // because a season nothing is known about absorbs NOTHING and « 0 épisodes
    // à récupérer » is both wrong in French, where zero takes the singular, and
    // a worse thing to read than saying so.
    //
    // AND A FOLLOW BEGUN BY THE ASK IS A SECOND FACT, so it has sentences of its
    // own: the same four, each with the follow said in it, chosen by the
    // answer's `newlyFollowed`. Never a clause appended to one of the others,
    // and never a key assembled from two halves: every key is written out, so
    // a reader searching for one finds the line that chooses it.
    const count = grab?.absorbedCount ?? 0;
    const newly = grab?.newlyFollowed === true;
    const messageKey = grab?.queued
      ? newly ? "seasonQueuedNewlyFollowed" : "seasonQueued"
      : count === 0
        ? newly ? "seasonAskedNoneNewlyFollowed" : "seasonAskedNone"
        : count === 1
          ? newly ? "seasonAskedOneNewlyFollowed" : "seasonAskedOne"
          : newly ? "seasonAskedNewlyFollowed" : "seasonAsked";
    window.__toast?.show({
      message: say(messageKey, { season, count }),
    });
    return grab?.queued === true;
  } catch {
    // THE REFUSAL IS SAID, and it is said as a refusal. Swallowing it would
    // leave the operator looking at a season that never moved with no reason
    // given — the silent failure this interface's own constitution refuses.
    window.__toast?.show({ message: say("seasonRefused", { season }) });
    return false;
  } finally {
    // IN A `finally`, so a refused ask can be made again. Released on the
    // throwing path as much as on the answering one: a guard that survived a
    // failure would make the second press — the one that follows a refusal and
    // is the operator's whole recourse — do nothing, for ever.
    inFlight.delete(asked);
    announceFlight();
  }
}
